python-sdk/nuscenes/tests/test_predict_helper.py

```diff
@@ -0,0 +1,503 @@
+                 samples: List[Dict[str, Any]]):
+
+        self._sample_annotation = {r['token']: r for r in sample_annotations}
+        self._sample = {r['token']: r for r in samples}
+
+    @property
+    def sample_annotation(self,) -> List[Dict[str, Any]]:
+        return list(self._sample_annotation.values())
+
+    def get(self, table_name: str, token: str) -> Dict[str, Any]:
+        assert table_name in {'sample_annotation', 'sample'}
+        return getattr(self, "_" + table_name)[token]
+
+class Test_convert_global_coords_to_local(unittest.TestCase):
+
+    def setUp(self):
+        along_pos_x = np.zeros((5, 2))
+        along_pos_y = np.zeros((5, 2))
+        along_neg_x = np.zeros((5, 2))
+        along_neg_y = np.zeros((5, 2))
+
+        along_pos_x[:, 0] = np.arange(1, 6)
+        along_pos_y[:, 1] = np.arange(1, 6)
+        along_neg_x[:, 0] = -np.arange(1, 6)
+        along_neg_y[:, 1] = -np.arange(1, 6)
+        self.along_pos_x, self.along_pos_y = along_pos_x, along_pos_y
+        self.along_neg_x, self.along_neg_y = along_neg_x, along_neg_y
+
+        y_equals_x = np.zeros((5, 2))
+        y_equals_x[:, 0] = np.arange(1, 6)
+        y_equals_x[:, 1] = np.arange(1, 6)
+        self.y_equals_x = y_equals_x
+
+    def test_heading_0(self):
+        rotation = (1, 0, 0, 0)
+        origin = (0, 0, 0)
+        offset = (50, 25, 0)
+
+        # Testing path along pos x direction
+        answer = convert_global_coords_to_local(self.along_pos_x, origin, rotation)
+        np.testing.assert_allclose(answer, self.along_pos_y, atol=1e-4)
+
+        answer = convert_global_coords_to_local(self.along_pos_x + [[50, 25]], offset, rotation)
+        np.testing.assert_allclose(answer, self.along_pos_y, atol=1e-4)
+
+        # Testing path along pos y direction
+        answer = convert_global_coords_to_local(self.along_pos_y, origin, rotation)
+        np.testing.assert_allclose(answer, self.along_neg_x, atol=1e-4)
+
+        answer = convert_global_coords_to_local(self.along_pos_y + [[50, 25]], offset, rotation)
+        np.testing.assert_allclose(answer, self.along_neg_x, atol=1e-4)
+
+        # Testing path along neg x direction
+        answer = convert_global_coords_to_local(self.along_neg_x, origin, rotation)
+        np.testing.assert_allclose(answer, self.along_neg_y, atol=1e-4)
+
+        answer = convert_global_coords_to_local(self.along_neg_x + [[50, 25]], offset, rotation)
+        np.testing.assert_allclose(answer, self.along_neg_y, atol=1e-4)
+
+        # Testing path along neg y direction
+        answer = convert_global_coords_to_local(self.along_neg_y, origin, rotation)
+        np.testing.assert_allclose(answer, self.along_pos_x, atol=1e-4)
+
+        answer = convert_global_coords_to_local(self.along_neg_y + [[50, 25]], offset, rotation)
+        np.testing.assert_allclose(answer, self.along_pos_x, atol=1e-4)
+
+    def test_heading_pi_over_4(self):
+        rotation = (np.cos(np.pi / 8), 0, 0, np.sin(np.pi / 8))
```

Comment:
The method names and the rotations are off by factor 2. Why is that?

Reply:
I'm using this convention `cos(theta / 2) + (xi + yi + zi)sin(theta / 2)` to convert between yaw and quaternion

Reply:
We usually use
- [quaternion_yaw](https://github.com/nutonomy/nuscenes-devkit/blob/053438468776a51f4a522cbf02023d54efdeaac3/python-sdk/nuscenes/eval/common/utils.py#L112)
- `q = Quaternion(axis=(0, 0, 1), angle=yaw_in)`